pytorch_lightning/trainer/connectors/signal_connector.py

```diff
@@ -61,7 +61,11 @@ def slurm_sigusr1_handler_fn(self, signum: Signals, frame: FrameType) -> None:
         if self.trainer.is_global_zero:
             # save weights
             log.info("handling SIGUSR1")
-            self.trainer.checkpoint_connector.hpc_save(self.trainer.weights_save_path, self.trainer.logger)
+            # save logger to make sure we get all the metrics
+            if self.trainer.logger:
+                self.trainer.logger.finalize("finished")
+            hpc_save_path = self.trainer.checkpoint_connector.hpc_save_path(self.trainer.weights_save_path)
+            self.trainer.save_checkpoint(hpc_save_path)
```

Comment:
this can't be under a `self.trainer.is_global_zero` check from L61. `trainer.save_checkpoint` relies on collectives, meaning it must be called on all ranks simultaneously. otherwise the user will see hangs in a distributed environment. 

To minimize changes in this PR, you could encapsulate just the logger finalization call under the `is_global_zero` check. `trainer.save_checkpoint` already takes care of all the other rank-specific checks. 

it looks like we don't have a test with deepspeed-3 and SLURM signal handling. in the code today, this would crash, because we need to save on all ranks in that case. 

so this is an awesome unification you're doing so we don't have to worry about divergent code paths like this anymore 😄

Reply:
looking at what's done in the main trainer, the logger finalization is called on all ranks, and it's up to logger implementations to determine if they only want to be called on rank 0 or not.

https://github.com/PyTorchLightning/pytorch-lightning/blob/7ed3dbf1913351868835ac3c549a0fbf6c612e09/pytorch_lightning/trainer/trainer.py#L1429-L1446

So I don't think we need to guard this with the global zero check either.

Reply:
I think* the second finalize call can be removed, but someone correct me if I'm missing something :)

Reply:
Hey @ananthsub. Great catch ! Yes, this should be covering only the requeing.

Yes, we should be call the logger finalize on all ranks, so let's remove the second one.